### PR TITLE
Launcher

### PR DESCRIPTION
**Pull request by vj-27**
issue #42 
todo : Windows

### CODE REVIEW

**Comment by domenukk** [comment]
At this point, the `launcher` function has so many arguments (even `clippy` warns about it), some of which are optional, that it may be time to use a builder?

**Comment by domenukk** [comment]
Before merging this, we should add all examples to ci (again)

**Comment by s1341** [comment]
Agreed

File: libafl/src/utils.rs [hunk 1]
```diff
@@ -655,7 +655,7 @@ where
                         self.shmem_provider.post_fork();
 
                         #[cfg(feature = "std")]
-                        std::thread::sleep(std::time::Duration::from_secs(id as u64));
+                        std::thread::sleep(std::time::Duration::from_secs((id + 1) as u64));
```
Comment (domenukk) [inline libafl/src/utils.rs:658]: Is this against shmem races or against broker <-> client races?

**Comment by domenukk** [comment]
I guess the b2b stuff should now become part of the restarting and launcher builders

File: libafl/src/bolts/shmem.rs [hunk 2]
```diff
@@ -209,7 +209,7 @@ pub trait ShMemProvider: Send + Clone + Default + Debug {
 #[derive(Debug, Clone)]
 pub struct RcShMem<T: ShMemProvider> {
     internal: ManuallyDrop<T::Mem>,
-    provider: Rc<RefCell<T>>,
+    provider: Rc<Mutex<RefCell<T>>>,
```
Comment (domenukk) [inline libafl/src/bolts/shmem.rs:212]: Erm... that means you need an `Arc`, too, right?
Where is this needed? For the b2b stuff? should we use a new process there, too, to get around this `Mutex`?

Comment (s1341) [reply libafl/src/bolts/shmem.rs:212]: I found that there were all sorts of races with the `RcShMem`, adding the `Mutex` resolved them. I can add an `Arc` if you think it's necessary.

File: libafl/src/utils.rs (hunk 1 above)
Comment (s1341) [reply libafl/src/utils.rs:658]: broker <-> client.

**Comment by domenukk** [comment]
this now only needs the individual fuzzer adaptation afaik. @andreafioraldi should we merge this with your other and make it v0.3 then?

**Comment by andreafioraldi** [comment]
> this now only needs the individual fuzzer adaptation afaik. @andreafioraldi should we merge this with your other and make it v0.3 then?

Yes the only thing left is fixing libafl_frida, you can directly rebase this branch over mine as I will not touch libafl

**Comment by andreafioraldi** [comment]
@domenukk is this pr ready? If so, tomorrow I will rebase on main, resolve the conflicts and merge. Then we should write a bit of doc about launcher and I have to update the book about the new feedback structure. After that, we can release.

**Comment by domenukk** [comment]
Libaflwise yes, the examples need fixing I guess. Also we should add them to ci...

**Comment by domenukk** [comment]
B2b may need some more testing (and probably some test cases, too)

**Comment by domenukk** [comment]
@andreafioraldi please double-check I merged your latest changes correctly, there were a lot of conflicts. Maybe next time we should try to merge long-running PRs before refactoring

**Comment by andreafioraldi** [comment]
No it's broken, libfuzzer_stb_image does not compile

**Comment by andreafioraldi** [comment]
All the stuffs in fuzzers/ are broken actually

**Comment by andreafioraldi** [comment]
`S: DeserializeOwned + IfInteresting<I, S>,` this line has no sense, IfInteresting is part of Fuzzer, and in this way is even a recursive generic with S: Something<S>

**Comment by andreafioraldi** [comment]
@domenukk I adapted libfuzzer_libpng_launcher to the new API, but it still doesn't work because you changes Launcher to a Builder and it uses the laucnher function. Beware that now fuzz_loop takes the stages as argument and you should change the return tuple type of the closure passed to the laucnher

**Comment by andreafioraldi** [comment]
Btw prob now that you are introducing builders we should create a Builder for State

**Comment by andreafioraldi** [comment]
And another small change before merging, maybe the launcher closure can take an index as parameter. In that way in the closure we can use different configuration, something like

```
if core_index == 0 {
    // enable CmpLog
}
```

**Comment by domenukk** [comment]
Pushed a CI script that at least tries to build them. Eventually, we need CI that runs the fuzzers for a while, too

**Comment by andreafioraldi** [comment]
The fuzzer works, but it is ugly. And, I still cannot use the cores at 100%, dunno why.

Screenshot at 2021-05-18 13-01-22

**Comment by s1341** [comment]
I manage to get 100% with almost all green on an android device.

**Review by andreafioraldi — APPROVED** [review]
> I manage to get 100% with almost all green on an android device.

**Comment by andreafioraldi** [comment]
> I manage to get 100% with almost all green on an android device.

Yes was my machine (13 days uptime with a fucked kernel setting), solved with a reboot.

I just figured out that I had an outdated pending review, don't read the last outdated comments.

**Comment by s1341** [comment]
let me check it tomorrow on android, and then we can merge...

**Comment by s1341** [comment]
@andreafioraldi Seems to work. There are still some bugs in android stuff, but I think they are unrelated to launcher. Go ahead and merge

**Comment by domenukk** [comment]
Good work everybody, thanks y'all ;)

**Comment by andreafioraldi** [comment]
I will update the book and release asap

**Pull request merged**